crates/ruff_python_formatter/src/expression/string.rs

```diff
@@ -262,129 +244,24 @@ impl Format<PyFormatContext<'_>> for FormatStringContinuation<'_> {
-            if fstring_range_builder.in_fstring() {
-                continue;
-            }
-
-            match token {
-                Tok::String { .. } | Tok::FStringEnd => {
-                    let token_range = if token.is_f_string_end() {
-                        fstring_range_builder.finish()
-                    } else {
-                        token_range
-                    };
-
-                    // ```python
-                    // (
-                    //      "a"
-                    //      # leading
-                    //      "the comment above"
-                    // )
-                    // ```
-                    let leading_comments_end = dangling_comments
-                        .partition_point(|comment| comment.start() <= token_range.start());
-
-                    let (leading_part_comments, rest) =
-                        dangling_comments.split_at(leading_comments_end);
-
-                    // ```python
-                    // (
-                    //      "a" # trailing comment
-                    //      "the comment above"
-                    // )
-                    // ```
-                    let trailing_comments_end = rest.partition_point(|comment| {
-                        comment.line_position().is_end_of_line()
-                            && !locator.contains_line_break(TextRange::new(
-                                token_range.end(),
-                                comment.start(),
-                            ))
-                    });
-
-                    let (trailing_part_comments, rest) = rest.split_at(trailing_comments_end);
-                    let part = StringPart::from_source(token_range, &locator);
-                    let normalized =
-                        part.normalize(self.string.quoting(&locator), &locator, quote_style);
-
-                    joiner.entry(&format_args![
-                        line_suffix_boundary(),
-                        leading_comments(leading_part_comments),
-                        normalized,
-                        trailing_comments(trailing_part_comments)
-                    ]);
-
-                    dangling_comments = rest;
-                }
-                Tok::Comment(_)
-                | Tok::NonLogicalNewline
-                | Tok::Newline
-                | Tok::Indent
-                | Tok::Dedent => continue,
-                token => unreachable!("Unexpected token {token:?}"),
-            }
+        for part in self.string.parts() {
+            let normalized = StringPart::from_source(part.range(), &locator).normalize(
+                self.string.quoting(&locator),
+                &locator,
+                quote_style,
+            );
+
+            joiner.entry(&format_args![
+                line_suffix_boundary(),
+                leading_comments(comments.leading(&part)),
+                normalized,
+                trailing_comments(comments.trailing(&part))
+            ]);
```

Comment:
I'm thinking of removing `FormatString` altogether and move the formatting logic in the respective node. The highlighted part is for implicit string concatenation and the `self.string.parts` is allocating a new vector because of type differences in each match blocks. This would also help in isolating docstring formatting only in `ExprStringLiteral`. Also, the f-string formatting would mean that the logic will be shifted so we might as well do it for string and bytes literal.

I'll probably do it in a follow-up PR.